kafka-rest/src/main/java/io/confluent/kafkarest/ProducerMetrics.java

```diff
@@ -23,40 +25,83 @@
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentMap;
 import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
 import org.apache.kafka.common.MetricName;
 import org.apache.kafka.common.MetricNameTemplate;
+import org.apache.kafka.common.metrics.CompoundStat;
 import org.apache.kafka.common.metrics.JmxReporter;
-import org.apache.kafka.common.metrics.KafkaMetricsContext;
 import org.apache.kafka.common.metrics.MeasurableStat;
 import org.apache.kafka.common.metrics.MetricConfig;
 import org.apache.kafka.common.metrics.Metrics;
 import org.apache.kafka.common.metrics.MetricsContext;
 import org.apache.kafka.common.metrics.MetricsReporter;
 import org.apache.kafka.common.metrics.Sensor;
 import org.apache.kafka.common.metrics.stats.Avg;
-import org.apache.kafka.common.metrics.stats.CumulativeCount;
-import org.apache.kafka.common.metrics.stats.CumulativeSum;
 import org.apache.kafka.common.metrics.stats.Max;
+import org.apache.kafka.common.metrics.stats.Percentile;
+import org.apache.kafka.common.metrics.stats.Percentiles;
 import org.apache.kafka.common.metrics.stats.Rate;
 import org.apache.kafka.common.metrics.stats.WindowedCount;
-import org.apache.kafka.common.metrics.stats.WindowedSum;
 import org.apache.kafka.common.utils.Time;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 public class ProducerMetrics {
-  public static final String JMX_PREFIX = "kafka.rest";
 
   private static final Logger log = LoggerFactory.getLogger(ProducerMetrics.class);
 
+  private final String fullyQualifiedRequestSensor;
+  private final String fullyQualifiedRequestSizeSensor;
+  private final String fullyQualifiedResponseSensor;
+  private final String fullyQualifiedResponseSizeSensor;
+  private final String fullyQualifiedRecordErrorSensor;
+  private final String fullyQualifiedRequestLatencySensor;
+
+  private final String jmxPrefix;
   private final Metrics metrics;
   private final ConcurrentMap<BeanCoordinate, ProduceMetricMBean> beansByCoordinate =
       new ConcurrentHashMap<>();
 
-  public ProducerMetrics(Time time) {
-    this.metrics = new MetricsBuilder(JMX_PREFIX).withTime(time).build();
+  public ProducerMetrics(KafkaRestConfig config, Time time) {
+    this.metrics = new MetricsBuilder(config.getMetricsContext()).withTime(time).build();
+    this.jmxPrefix = config.getString(KafkaRestConfig.METRICS_JMX_PREFIX_CONFIG);
+    this.fullyQualifiedRecordErrorSensor =
+        String.join(
+            ":",
+            jmxPrefix,
+            ProducerMetricsRegistry.GROUP_NAME,
+            ProducerMetricsRegistry.RECORD_ERROR_SENSOR);
```

Comment:
I think this is more efficient as we only compute these strings once on ProducerMetrics creation. Given the overhead for creating templates this is more efficient than String.format too.

For readability I think the separate convenience methods for recording metrics are more readable than a templating solution. If we had a lot of them it would become unmanageable but given there are only 5 i think we are ok.

Reply:
I think we might be disagreeing because we are talking about different things :)

My suggestion is not to remove the actual precomputation, as that was the main optimization from the previous version - just to replace the `String.join` calls here with either a `String.format` with a single parameter, or even directly concatenating the common prefix/template with the actual sensor name. There is no meaningful overhead in creating what I call template above - it is a single `String` which won't escape YoungGen as it is needed solely in the constructor.

The separate convenience methods won't be affected at all.

Reply:
ah, i'm with you and have changed.